src/Cardano/Shell/Lib.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Shell.Lib
+    ( ApplicationEnvironment (..)
+    , GeneralException (..)
+    , CardanoApplication (..)
+    , initializeAllFeatures
+    , runCardanoApplicationWithFeatures
+    -- * configuration for running
+    , loadCardanoConfiguration
+    , initializeCardanoEnvironment
+    ) where
+
+import Cardano.Prelude
+
+import Cardano.Shell.Types
+
+import Cardano.Shell.Features.Logging
+import Cardano.Shell.Features.Networking
+
+--------------------------------------------------------------------------------
+-- General exceptions
+--------------------------------------------------------------------------------
+
+data GeneralException
+    = UnknownFailureException -- the "catch-all"
+    | MissingResourceException
+    | FileNotFoundException
+    deriving (Eq, Show)
+
+instance Exception GeneralException
```

Comment:
Do we really want to use exceptions? Especially since the constructors `MissingResourceException` and `FileNotFoundException` do not sound much like exceptions. They are flow control.

Reply:
At this point, they are just examples of some things that can go wrong that can't be handled from the actual module/feature. 

> They are flow control.

Well, when a module/feature doesn't find a file that it requires for running, what is the result of that control flow? I presume termination - so the point here is to have a top-level information about that termination and why it happened and restart it.

In any case, we don't have actual examples of how we could use this, it's just my idea of what would be a nice general way to inform the top level application of what went wrong.